### PR TITLE
expand path of systemctl link argument

### PR DESCRIPTION
**Pull request by boucman**
systemctl link is the only systemctl verb that takes a filename (and not
a unit name) as argument

use realpath to expand the provided filename in order to make it easier
to use from the command line

### CODE REVIEW

File: src/systemctl/systemctl.c [hunk 1]
```diff
@@ -6001,6 +6001,24 @@ static int mangle_names(char **original_names, char ***mangled_names) {
         return 0;
 }
 
+static int normalize_filenames(char **names, bool warn_if_path) {
+        char **u;
+
+        STRV_FOREACH(u, names) {
+                char* normalized_path;
+
+                normalized_path = realpath(*u,NULL);
```
Comment (keszybz) [inline src/systemctl/systemctl.c]: Space after comma, see CODING_STYLE.

Comment (boucman) [reply src/systemctl/systemctl.c]: ok, will do

Comment (boucman) [reply src/systemctl/systemctl.c]: done

Comment (keszybz) [reply src/systemctl/systemctl.c]: Did you forget to commit your changes?

**Comment by keszybz** [comment]
@boucman something's wrong with your patches — it looks like you didn't actually update the commits. Please check.

**Comment by poettering** [comment]
realpath() is not an option btw, as that cannot take --root= into account. We have the chase_symlinks stuff for that case.

Also, note that "systemctl enable" also accepts absolute paths in addition to plain unit names. And for those we shouldn't follow symlinks either (as that would mean following the config we are supposed to create). Hence I am really not convinced that "systemctl link" should follow symlinks at all.

**Comment by boucman** [comment]
@keszybz : my bad, pushed again

hmm I kinda expectes realpath() to not be Ok but I hadn't found an alternative, I'll look into chase_symlink.

note that my patch only acted on the "link" verb because its argument is always a filename. I am not sure what is the intended use-case for "systemctl link" but i use it mainly to debug a service I am writing, and having to type "systemctl link $(pwd)/my.service" is a bit annoying. 

I'd really like to have filenanmes handled in the more usuall way by systemd (especially systemd-analyze verify and systemctl link) if you tell me what you want I can code it, would a "lower priority" filename expansion for systemctl enable help ? or are you strongly oposed to expanding filenames to path ?

i'm ok to rework, just tell me what to do...

File: src/systemctl/systemctl.c [hunk 2]
```diff
@@ -6001,6 +6001,24 @@ static int mangle_names(char **original_names, char ***mangled_names) {
         return 0;
 }
 
+static int normalize_filenames(char **names, bool warn_if_path) {
```
Comment (keszybz) [inline src/systemctl/systemctl.c]: Why do you pass `warn_if_path` here?

**Comment by keszybz** [comment]
I don't think `enable` is that relevant here. `link` is a much simpler thing, and it can be more liberal in the arguments it accepts because there is less ambiguity.

However, with `--root` things get more complicated.

`systemctl link --root= /a/b.service` → this operates on `/a/b.service`
`systemctl link --root=/some/path /a/b.service` → this operates on `/some/path/a/b.service`
(so far so good)
`systemctl link --root= ./b.service` → this operates (with your patch) on `$PWD/b.service`
`systemctl link --root=/some/path ./b.service` → problem

I'd say that relative paths are fundamentally incompatible with `--root` and should be disallowed.

File: src/systemctl/systemctl.c (hunk 2 above)
Comment (boucman) [reply src/systemctl/systemctl.c]: because i'm stupid...

I started by copy/pasting normalize_name and missed the unused parameter. fixing...

**Comment by boucman** [comment]
I can check that arg_root is empty, no problem. I'll resubmit in a bit, in case other comments trickle in...

**Comment by poettering** [comment]
@boucman so, what's you usecase here? as I understand you your interest is in resolving relative paths to absolute paths, no? It's not so much in having the tool follow symlinks for you, is it?

If so, we should really just qualify the relative path and not use realpath() or chase_symlinks() at all...

**Comment by boucman** [comment]
it's more a case of expecting "systemctl link file_in_cwd.service" to create a link to a file in the current working directory, as most CLI software do. 

it's just a case of ease of use. currenly I have to type $CWD/file_in_cwd.service which is not very comfortable and auto-completion doesn't work....

so yes, qualifying the relative path would be enough for me, but realpath() was the only call i could find doing that. Maybe I miss something, i'll look in basic/*.h some more.

**Comment by keszybz** [comment]
@poettering I think `realpath` is good enough here, since if --root is disallowed, symlinks can be followed in the usual way.

**Comment by poettering** [comment]
> @poettering I think realpath is good enough here, since if --root is disallowed, symlinks can be followed in the usual way.

Well, but the symlink stuff is not what @boucman is after... Also, chase_symlinks() is fully equivalent to realpath() afaics and it does understand the root dir...

**Comment by poettering** [comment]
> so yes, qualifying the relative path would be enough for me, but realpath() was the only call i could find doing that. Maybe I miss something, i'll look in basic/*.h some more.

You are looking for path_make_absolute_cwd()

**Comment by poettering** [comment]
> so yes, qualifying the relative path would be enough for me, but realpath() was the only call i could find doing that. Maybe I miss something, i'll look in basic/*.h some more.

You are looking for path_make_absolute_cwd().

**Comment by boucman** [comment]
@poettering Not 100% sure that's what I want... 
if I pass a path like ~/tmp/bla.service, this doesn't look like it would expand properly...

if I pass ../../../whatever/test.service the result would not be a "simplified" path, but $CWD/../../whatever/test.service

I am not sure exactly what would be the expected result here... In my case (testing) that's good enough, but i'm afraid that, long term, this is not what people would expect to happen

In a way, I don't really care about symlinks because symlinks are transparent to the human-user, but I do care about ~, .. and other path-expansion mechanism. So far chase_symlinks seems a good candidate, but, as I said, I'm happy to discuss use-cases.

**Comment by keszybz** [comment]
@boucman: ~ is expanded by the shell, so path_make_absolute_cwd will do what you need.

**Comment by keszybz** [comment]
If you could update the patch today, that'd be great. I'd prefer to have this in v234, and this is one of the few PRs left.

**Comment by boucman** [comment]
sorry for the delay, i'm a "week-end coder" on that one... will update today

**Comment by boucman** [comment]
Ok, I pushed a new (and much simpler) version

I only expand path if there is no --root; I will look into dealing with --root (using chase_symlink) in a separate PR once I have something working

**Comment by keszybz** [comment]
So... this is almost good, but I have two requests:
1. add a comment in that `if` that non-absolute paths will be rejected later
2. check that none of the paths are single-component paths, i.e. that that all contain at least a single slash.

For 2., I think adding a `bool` argument to `path_strv_make_absolute_cwd` would be the easiest option. It's only used in one other place.

**Comment by boucman** [comment]
no problem with 1. but i'm not sure I understand 2.
you want me to reject "systemctl my_service.service" ? 

that's the point of the PR, so i'm a bit suprised...

**Comment by keszybz** [comment]
Yes, exactly. I think `./whatever.service` should be allowed, but not `whatever.service`.

**Comment by boucman** [comment]
hmm, I disagree on the ground of "that's not how most CLI programs work" so it kinda suprises the user... 

i'll implement if you want to, but if you could give me a little more rational, i'd apreciate that...

**Comment by keszybz** [comment]
I did that in my first comment, I think ;)

Essentially, "./foobar" is not much more work to type than "foobar", but avoid any potential confusion.

**Comment by poettering** [comment]
> Yes, exactly. I think ./whatever.service should be allowed, but not whatever.service.

Ah, I like this idea to complain. But I wonder if just warning would be better than to refuse? Dunno? Fine with eiher...

**Comment by boucman** [comment]
Ok, i'll rework to check that this WE

**Comment by boucman** [comment]
I there a use-case for "systemctl link \<unit\>" (as opposed to file) ?

I don't really see one, so systemctl link bla.service is non-ambiguous since we can't mean the unit, only a file bla.service in the current directory...

maybe I should disable normal unit-name expansion for the link case altogether...

**Comment by boucman** [comment]
ok, updated to check for '/'

please review

**Comment by keszybz** [comment]
> I there a use-case for "systemctl link <unit>" (as opposed to file)?

There isn't. Maybe I'm being overly cautious. But for my defence, it's easier to relax the requirement than to add it later.

**Comment by boucman** [comment]
fair enough, let's do it that way, then.

**Comment by keszybz** [comment]
I mangled your commit a bit and force-pushed to your branch, I hope that's OK. Now the error message is a bit more explicit and gives the reason for refusal.

If people are annoyed by the constraint on pathnames, we can always relax it later. But let's give it some time in the wild.

**Comment by boucman** [comment]
sure, looks better that way

**Comment by keszybz** [comment]
`× xenial-s390x`
badpkg: Test dependencies are unsatisfiable.
`o xenial-i386` seems stuck.

**Pull request merged**